### ISSUE TITLE
Missing type-hinting in Mock object of abstract class with type-hinting in method signature

### ISSUE THREAD

**Issue by beToiba** [sebastianbergmann/phpunit-mock-objects#106]
Class under test

``` php
abstract class Chef
{
    public function prepareIngridients()
    {
        // calls implementation of cut
    }

    /**
     * @param Vegetable $ingridient
     */
    protected abstract function cut(Vegetable $ingridient);
    /* do other things */
}
```

Test

``` php
class ClassChefTest extends PHPUnit_Framework_TestCase
{
    public function test()
    {
        $stub = $this->getMockBuilder('Chef')
                               ->setMethods(ArrayOfMethodsToMock)
                               ->getMockForAbstractClass();
        //some cool testing stuff ...
    }
}
```

Output:

> [...]
> PHPUnit 3.7.1 by Sebastian Bergmann.
> [...]
> PHP Fatal error:  Declaration of Mock_Chef_53c223e3::cut() must be compatible with that of Chef::cut() in /usr/share/php/PHPUnit/Framework/MockObject/Generator.php(231) : eval()'d code on line 1
> [StackTrace...]
> eval(''...

eval has something like this in it:

``` php
protected function cut($ingridient) // <--- Type hinting is missing
    {
        $arguments = array($ingridient);
        $count     = func_num_args();        
        if ($count > 1) {
            $_arguments = func_get_args();
            for ($i = 1; $i < $count; $i++) {
                $arguments[] = $_arguments[$i];
            }
        }

        $result = $this->__phpunit_getInvocationMocker()->invoke(
            new PHPUnit_Framework_MockObject_Invocation_Object(
                'Chef', 'cut', $arguments, $this, FALSE          
            )
        );
        return $result;
    }
```

**Comment by MacDada** [comment]
> This usually happens when the type-hinted class is not loaded at the time the mock is generated.

Hi, I stumbled upon this problem. It used to work when I wrote the test, but now I opened an old project and it doesn't work anymore.

The problem:

1. A method contains typehint to a framework's class
2. I mock that method.
3. The framework is **not** loaded when running the tests (and cannot be, reaaaly legacy project)
4. `Strict Standards: Declaration of Mock_Foo_3a1fc1ab::bar() should be compatible with Foo::Bar(Baz $user) in vendor/phpunit/phpunit-mock-objects/PHPUnit/Framework/MockObject/Generator.php(231) : eval()'d code on line 1`

* I don't want to remove the typehint.
* I don't want to load the framework in the tests

Any solutuon? PHPUnit 3.7.38
